assert/assertions.go

```diff
@@ -2292,3 +2301,357 @@ func buildErrorChainString(err error, withType bool) string {
+// in the same order, with the same number of occurrences of each element.
+// Every element in the expected array must be present in the actual array
+// at the same index, and vice versa. It also performs deep equality checks
+// on each element's fields.
+//
+//	assert.Match(t, []Employee{e1, e2}, []Employee{e1, e2})
+func Match(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{}) bool {
+	if h, ok := t.(tHelper); ok {
+		h.Helper()
+	}
+
+	if !isSliceOrArray(expected) || !isSliceOrArray(actual) {
+		return Fail(t, "Match function works only with slices or arrays", msgAndArgs...)
+	}
+
+	return inspectSlices(t, expected, actual, msgAndArgs...)
+}
+
+// isSliceOrArray checks if the provided value is a slice or an array.
+// It handles nil values by returning false.
+func isSliceOrArray(v interface{}) bool {
+	if v == nil {
+		return false
+	}
+	kind := reflect.TypeOf(v).Kind()
+	return kind == reflect.Slice || kind == reflect.Array
+}
+
+// inspectSlices compares each element of two slices or arrays for deep equality.
+// It provides detailed error messages when differences are found, including the specific fields that differ.
+func inspectSlices(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool {
+	expectedValue := reflect.ValueOf(expected)
+	actualValue := reflect.ValueOf(actual)
+
+	expectedLen := expectedValue.Len()
+	actualLen := actualValue.Len()
+
+	minLen := expectedLen
+	if actualLen < minLen {
+		minLen = actualLen
+	}
+
+	var differences []string
+
+	if actualLen != expectedLen {
+		Fail(t, fmt.Sprintf("Lengths not equal:\nexpected: %d\nactual  : %d", expectedLen, actualLen), msgAndArgs...)
+		return false
+	}
+
+	for i := 0; i < minLen; i++ {
+		expectedItem := expectedValue.Index(i).Interface()
+		actualItem := actualValue.Index(i).Interface()
+
+		if !ObjectsAreEqual(expectedItem, actualItem) {
+			message := messageFromMsgAndArgs(msgAndArgs...)
+			diffs := findDifferences(expectedItem, actualItem)
+
+			differencesOutput := "Field differences:\n"
+			for _, diff := range diffs {
+				differencesOutput += fmt.Sprintf("  └─ %s: %s ≠ %s\n", diff.Path, formatDiffValue(diff.Expected), formatDiffValue(diff.Actual))
+			}
+
+			if message != "" {
+				differences = append(differences, fmt.Sprintf(
+					"MATCH: %s\n[index %d] Not equal:\nexpected: %v\nactual  : %v",
+					message, i, formatComparisonValue(expectedItem), formatComparisonValue(actualItem)))
+				differences = append(differences, differencesOutput)
+			} else {
+				differences = append(differences, fmt.Sprintf(
+					"MATCH: [index %d] Not equal:\nexpected: %v\nactual  : %v",
+					i, formatComparisonValue(expectedItem), formatComparisonValue(actualItem)))
+				differences = append(differences, differencesOutput)
+			}
+		}
+	}
+
+	if len(differences) > 0 {
+		diffMessage := strings.Join(differences, "\n")
+		Fail(t, fmt.Sprintf("Differences found:\n%s", diffMessage), msgAndArgs...)
+		return false
+	}
+
+	return true
```

Comment:
This might improve the readability

```suggestion
	if len(differences) == 0 {
		return true
	}

	diffMessage := strings.Join(differences, "\n")
	return Fail(t, fmt.Sprintf("Differences found:\n%s", diffMessage), msgAndArgs...)
```